x/mongo/driver/topology/topology.go

```diff
@@ -160,9 +151,34 @@ func (t *Topology) Connect() error {
 	t.desc.Store(description.Topology{})
 	var err error
 	t.serversLock.Lock()
+
+	// A replica set name sets the initial topology type to ReplicaSetNoPrimary unless a direct connection is also
+	// specified, in which case the initial type is Single.
+	if t.cfg.replicaSetName != "" {
+		t.fsm.SetName = t.cfg.replicaSetName
+		t.fsm.Kind = description.ReplicaSetNoPrimary
+	}
+
+	// A direct connection unconditionally sets the topology type to Single.
+	if t.cfg.mode == SingleMode {
+		t.fsm.Kind = description.Single
+	}
+
+	for _, a := range t.cfg.seedList {
+		addr := address.Address(a).Canonicalize()
+		t.fsm.Servers = append(t.fsm.Servers, description.NewDefaultServer(addr))
+	}
+
+	// store new description
+	newDesc := description.Topology{
+		Kind:                  t.fsm.Kind,
+		Servers:               t.fsm.Servers,
+		SessionTimeoutMinutes: t.fsm.SessionTimeoutMinutes,
+	}
+	t.desc.Store(newDesc)
+	t.publishTopologyDescriptionChangedEvent(description.Topology{}, t.fsm.Topology)
```

Comment:
Since this occurs while `serversLock` is held, I assume if a user callback did anything requiring server selection on the same client. This is inconsistent behavior among drivers. Callbacks for libmongoc similarly would deadlock. If that is the case, I suggest noting in the docs for the callbacks that the callbacks should not attempt to do any operation on the same client requiring server selection. See also: https://jira.mongodb.org/browse/DRIVERS-1277